### PR TITLE
[ir] Refactor result fetching

### PR DESCRIPTION
**Pull request by lin-hitonami**
This PR adds code that fetches the return values out of a StructType instead of uint64 arrays. The next PR will let the kernel return a StructType instead of a uint64 array.

Note that before the refactoring, f16 was converted to f32 before returning. However after the refactor, we convert the f16 to f32 on host after it is returned (in the next PR). Therefore, I added an implementation of f16-to-f32 conversion (from stackoverflow) to util/bits.cpp.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #7310
* __->__ #7262
* #7261

### CODE REVIEW

File: taichi/runtime/llvm/llvm_runtime_executor.h [hunk 1]
```diff
@@ -85,6 +85,15 @@ class LlvmRuntimeExecutor {
         fetch_result_uint64(i, result_buffer));
```
Comment (bobcao3) [inline taichi/runtime/llvm/llvm_runtime_executor.h:85]: We can finally remove this API afterwards right? `fetch_result_uint64` also crept into `device.h` and should be removed imo

**Review by bobcao3 — APPROVED** [review]
LGTM!

File: taichi/runtime/llvm/llvm_runtime_executor.h (hunk 1 above)
Comment (lin-hitonami) [reply taichi/runtime/llvm/llvm_runtime_executor.h:85]: Yeah, after the refactoring of all backends completes.

**Review by strongoier — APPROVED** [review]

**Pull request merged**